src/eval/README.md

```diff
@@ -1,82 +1,107 @@
+   layout store so subsequent lookups are cheap.
+6. **Polymorphic calls** – Before a function call, `prepareCall` consults the
+   poly cache. The interpreter only re-runs the runtime unifier if it has not
+   seen that combination of function id + argument Vars before.
+7. **Crash handling** – Crash/expect expressions delegate to the host via
+   `RocOps.crash`. Hosts supply a `CrashContext` (see `crash_context.zig`) to
+   record messages; the interpreter keeps no internal crash state.
+
+All RocOps interactions (alloc, dealloc, crash, expect) happen through the
+`RocOps` pointer passed into `evalMinimal`. This keeps host integrations (REPL,
+snapshot tool, CLI) consistent.
+
+## Rendering
+
+`renderValueRoc` and `renderValueRocWithType` assemble human-readable strings
+using the same type information the interpreter evaluated with. Rendering only
+reads from `StackValue` and the runtime layout store, so callers should decref
+the evaluated value *after* rendering.
+
+## Extending the Interpreter
+
+- **New expression forms** – Add cases to `evalExprMinimal`. Most cases follow a
+  pattern: translate sub-expressions, obtain or build layouts, then use the
+  helpers in `StackValue` to initialize the result.
+- **New data shapes** – Extend layout translation in
+  `translateTypeVar`/`getRuntimeLayout` and teach `StackValue` how to copy or
+  decref the shape.
+- **Rendering** – Update `render_helpers.zig` and ensure the interpreter calls
+  the appropriate helper.
+
+When making changes, run `zig build test`. Interpreter-specific coverage lives
+in:
+
+- `src/eval/test/interpreter_style_test.zig` – End-to-end Roc-syntax tests that
+  parse, canonicalize, evaluate, and render.
+- `src/eval/test/interpreter_polymorphism_test.zig` – Scenarios that exercise
+  the polymorphism cache and runtime unifier.
+- `src/repl/repl_test.zig` – Integration-style tests that ensure the REPL uses
+  the interpreter correctly.
+
+## Host Integrations
+
+- **REPL** (`src/repl/Repl.zig`) constructs a fresh interpreter per evaluation,
+  feeds it a canonical expression, then renders values through the interpreter’s
+  helpers.
+- **Snapshot tool** (`src/snapshot_tool/main.zig`) uses the same interpreter to
+  evaluate each snapshot input with optional tracing.
+- **Interpreter shim** (`src/interpreter_shim/main.zig`) provides a C-callable
+  entry point that deserializes a `ModuleEnv`, constructs an interpreter, and
+  returns rendered output.
+
+## Tips for Contributors
+
+- Use the provided helpers (`StackValue.copyToPtr`, `StackValue.decref`, render
+  functions) instead of manipulating raw pointers—this keeps refcounting
+  correct.
+- The runtime stores (`runtime_types`, `runtime_layout_store`) are owned by the
+  interpreter instance. Reuse the same interpreter when evaluating multiple
+  expressions inside a single host context so caches pay off.
+- When debugging type translation, the `tests/interpreter_*` suites have targeted
```

Comment:
I don't think we have any tests at `tests/interpreter_*` this may need editing a little.